madmom/ml/nn/layers.py

```diff
@@ -945,3 +945,87 @@ def activate(self, data, **kwargs):
         return self.activation_fn(
             (data - self.mean) * (self.gamma * self.inv_std) + self.beta
         )
+
+
+class FunctionLayer(Layer):
+    """
+    Generic layer that applies a given function to the data.
+
+    Parameters
+    ----------
+    func : callable
+        Function to be applied to the data.
+    *args : positional arguments
+        Positional arguments to be passed to `func`.
+    **kwargs : keyword arguments
+        Keyword arguments to be passed to `func`.
+
+    """
+
+    def __init__(self, func, *args, **kwargs):
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+
+    def activate(self, data, **kwargs):
+        """
+        Activate the layer.
+
+        Parameters
+        ----------
+        data : numpy array
+            Activate with this data.
+
+        Returns
+        -------
+        numpy array
+            Activations for this data.
+
+        """
+        return self.func(data, *self.args, **self.kwargs)
+
+
+class PadLayer(Layer):
+    """
+    Padding layer that pads the input with a constant value.
+ 
+    Parameters
+    ----------
+    width : int
+        Width of the padding (only one value for all dimensions)
+    axes : iterable
+        Indices of axes to be padded
+    value : float
+        Value to be used for padding.
+
+    """
+
+    def __init__(self, width, axes, value=0.):
+        self.width = width
+        self.axes = axes
+        self.value = value
+
+    def activate(self, data, **kwargs):
+        """
+        Activate the layer.
+
+        Parameters
+        ----------
+        data : numpy array
+            Activate with this data.
+
+        Returns
+        -------
+        numpy array
+            Activations for this data.
+
+        """
+        shape = list(data.shape)
+        data_idxs = [slice(None) for _ in range(len(shape))]
+        for a in self.axes:
+            shape[a] += self.width * 2
+            data_idxs[a] = slice(self.width, -self.width)
+        data_padded = np.empty(tuple(shape))
+        data_padded[:] = self.value
+        data_padded[data_idxs] = data
+        return data_padded
```

Comment:
First of all, I'd like to see the NN changes split into a separate commit.
The layer itself looks good to me, however it would be great to unify [with the padding applied for the CNNOnsetProcessor](https://github.com/CPJKU/madmom/blob/master/madmom/features/onsets.py#L802) which needs repeating of the first/last frame N times.

Reply:
Hm, ok, then repeating is necessary, after all :)

Do you mean a separate commit or a separate PR? Anyways, I agree that the `PadLayer` could be more flexible, but I still think we can merge it with a bare minimum of functionality and add more afterwards.

Reply:
I find it weird that the commit message of adding a PaddingLayer is "implemented key recognition feature", thus I think this should be a separate commit. There's no need for a separate PR and we can move the padding of `CNNOnsetDetector` later on, but we should rather think about argument names and needed functionality now than later.